paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc

```diff
@@ -61,414 +61,425 @@ static mkldnn::memory::data_type GetDstType(bool is_int8, bool is_bfloat16,
+  const int groups = std::max(ctx.Attr<int>("groups"), 1);
+
+  const auto& scale_weights_data =
+      ctx.Attr<std::vector<float>>("Scale_weights");
+  const auto& scale_in_data = ctx.Attr<float>("Scale_in");
+
+  bool is_multi_channel = scale_weights_data.size() > 1;
+  int mask_reorder = is_multi_channel ? 1 << 0 : 1;
+  const int count =
+      is_multi_channel
+          ? (groups > 1 ? (weights_tz)[1] * (weights_tz)[0] : (weights_tz)[0])
+          : 1;
+
+  bias_scale_tuple =
+      std::make_shared<std::tuple<float, std::vector<float>>>(std::make_tuple(
+          static_cast<float>(mask_reorder), std::vector<float>(count)));
+  for (int i = 0; i < count; i++) {
+    std::get<1>(*bias_scale_tuple)[i] = scale_in_data * scale_weights_data[i];
+  }
+
+  dev_ctx.SetBlob(key_bs, bias_scale_tuple);
+
+  return bias_scale_tuple;
+}
+
+template <typename T>
+std::shared_ptr<std::tuple<float, std::vector<float>>> get_scales(
+    const framework::ExecutionContext& ctx,
+    const platform::MKLDNNDeviceContext& dev_ctx, const std::string& key) {
+  return std::make_shared<std::tuple<float, std::vector<float>>>(
+      std::make_tuple(1.0f, std::vector<float>()));
+}
+
+template <>
+std::shared_ptr<std::tuple<float, std::vector<float>>> get_scales<int8_t>(
+    const framework::ExecutionContext& ctx,
+    const platform::MKLDNNDeviceContext& dev_ctx, const std::string& key) {
+  // Get scales int8 bias key
+  const std::string key_s = key + "@s";
+
+  // Scales for int8 bias are to be cached to avoid
+  // computing them each iteration
+  auto scale_tuple =
+      std::static_pointer_cast<std::tuple<float, std::vector<float>>>(
+          dev_ctx.GetBlob(key_s));
+  if (scale_tuple) return scale_tuple;
+
+  const auto* filter = ctx.Input<Tensor>("Filter");
+  const auto& weights_tz = framework::vectorize(filter->dims());
+
+  const bool& force_fp32_output = ctx.Attr<bool>("force_fp32_output");
+  const bool& fuse_residual_conn = ctx.Attr<bool>("fuse_residual_connection");
+  const int groups = std::max(ctx.Attr<int>("groups"), 1);
+
+  const auto& scale_in_data = ctx.Attr<float>("Scale_in");
+  const auto& scale_in_eltwise_data = ctx.Attr<float>("Scale_in_eltwise");
+  auto scale_weights_data = ctx.Attr<std::vector<float>>("Scale_weights");
+  bool is_multi_channel = scale_weights_data.size() > 1;
+  auto scale_out_data = force_fp32_output ? 1.0f : ctx.Attr<float>("Scale_out");
+  float sum_scale =
+      fuse_residual_conn ? scale_out_data / scale_in_eltwise_data : 1.0f;
+  int count = is_multi_channel ? (groups > 1 ? (weights_tz)[1] * (weights_tz)[0]
+                                             : (weights_tz)[0])
+                               : 1;
+  std::vector<float> output_shift_scale(count);
+
+  scale_tuple =
+      std::make_shared<std::tuple<float, std::vector<float>>>(std::make_tuple(
+          static_cast<float>(sum_scale), std::vector<float>(count)));
+  for (int i = 0; i < count; i++) {
```

Comment:
for loop could use #progma omp ?

Reply:
This generation of scales is performed only once as it is cached now. So there is no need adding openmp as it would just create multiple threads . each thread with some resources that would just be not used later on. So that is why I do not add omp parallel here

Reply:
Ok ~